File: android/app/src/main/java/com/mobilelivecaster/streaming/NativeAdaptiveBitrateController.kt
package com.mobilelivecaster.streaming

import kotlin.math.roundToInt

internal data class NativeAdaptiveBitrateSample(
    val nowElapsedMs: Long,
    val nowWallMs: Long,
    val active: Boolean,
    val publishGeneration: Int,
    val congested: Boolean,
    val queuedItems: Int,
    val cacheSize: Int,
    val measuredBitrateKbps: Int,
    val useMeasuredBitrate: Boolean,
    val droppedVideoFrames: Long,
    val cumulativeReconnectCount: Int,
    val thermalState: String = "unknown"
)

internal data class NativeAdaptiveBitrateDecision(
    val type: String,
    val targetKbps: Int,
    val reason: String
)

internal data class NativeAdaptiveBitrateSnapshot(
    val controlOwner: String = "native",
    val controllerState: String = "idle",
    val baselineTargetKbps: Int = 0,
    val effectiveTargetKbps: Int = 0,
    val floorTargetKbps: Int = 0,
    val pendingTargetKbps: Int = 0,
    val automaticReductionCount: Long = 0,
    val automaticRestorationCount: Long = 0,
    val pressureSampleCount: Int = 0,
    val healthySampleCount: Int = 0,
    val cooldownRemainingMs: Long = 0,
    val recoveryEligibleInMs: Long = 0,
    val publishGeneration: Int = 0,
    val cumulativeReconnectCount: Int = 0,
    val lastDecisionAt: Long = 0,
    val lastDecisionReason: String = ""
)

internal fun isNativeAdaptiveBitrateActive(
    sessionLive: Boolean,
    publishGeneration: Int,
    publisherState: String,
    transportActive: Boolean
): Boolean = sessionLive && publishGeneration > 0 && publisherState == "published" && transportActive

internal fun resolveNativeVideoBitrateTargetKbps(
    requestedTargetKbps: Int,
    baselineChanged: Boolean,
    controller: NativeAdaptiveBitrateSnapshot
): Int = controller.pendingTargetKbps.takeIf { it > 0 }
    ?: if (baselineChanged) requestedTargetKbps else controller.effectiveTargetKbps.takeIf { it > 0 }
    ?: requestedTargetKbps

internal class NativeAdaptiveBitrateController {
    companion object {
        const val STARTUP_GRACE_MS = 10_000L
        const val REDUCTION_COOLDOWN_MS = 8_000L
        const val RECOVERY_COOLDOWN_MS = 20_000L
        const val RECOVERY_AFTER_REPUBLISH_MS = 30_000L
        const val PRESSURE_SAMPLES = 3
        const val CRITICAL_PRESSURE_SAMPLES = 2
        const val HEALTHY_SAMPLES = 30
        const val ABSOLUTE_MINIMUM_KBPS = 900
        const val MINIMUM_BASELINE_PERCENT = 35
    }

    private var controllerState = "idle"
    private var baselineTargetKbps = 0
    private var effectiveTargetKbps = 0
    private var floorTargetKbps = 0
    private var pendingTargetKbps = 0
    private var pendingBaselineTargetKbps: Int? = null
    private var pendingDecisionType: String? = null
    private var automaticReductionCount = 0L
    private var automaticRestorationCount = 0L
    private var pressureSampleCount = 0
    private var criticalPressureSampleCount = 0
    private var healthySampleCount = 0
    private var cooldownUntilElapsedMs = 0L
    private var recoveryEligibleElapsedMs = 0L
    private var publishedAtElapsedMs = 0L
    private var publishGeneration = 0
    private var cumulativeReconnectCount = 0
    private var reconnectPressurePending = false
    private var lastDroppedVideoFrames = 0L
    private var lastDecisionAt = 0L
    private var lastDecisionReason = ""

    fun reset(baselineKbps: Int, nowElapsedMs: Long = 0L) {
        baselineTargetKbps = normalizeTarget(baselineKbps)
        effectiveTargetKbps = baselineTargetKbps
        floorTargetKbps = floorForBaseline(baselineTargetKbps)
        pendingTargetKbps = 0
        pendingBaselineTargetKbps = null
        pendingDecisionType = null
        automaticReductionCount = 0
        automaticRestorationCount = 0
        pressureSampleCount = 0
        criticalPressureSampleCount = 0
        healthySampleCount = 0
        cooldownUntilElapsedMs = 0
        recoveryEligibleElapsedMs = nowElapsedMs.coerceAtLeast(0L)
        publishedAtElapsedMs = 0
        publishGeneration = 0
        cumulativeReconnectCount = 0
        reconnectPressurePending = false
        lastDroppedVideoFrames = 0
        lastDecisionAt = 0
        lastDecisionReason = ""
        controllerState = "idle"
    }

    fun requestBaselineChange(baselineKbps: Int, thermalState: String = "unknown"): Int {
        val target = normalizeTarget(baselineKbps)
        val normalizedThermalState = thermalState.trim().lowercase()
        val currentTarget = listOfNotNull(
            effectiveTargetKbps.takeIf { it > 0 },
            pendingTargetKbps.takeIf { it > 0 }
        ).minOrNull() ?: target
        pendingTargetKbps = when (normalizedThermalState) {
            "critical" -> minOf(floorForBaseline(target), currentTarget)
            "fair", "serious" -> minOf(target, currentTarget)
            else -> target
        }
        pendingBaselineTargetKbps = target
        pendingDecisionType = null
        reconnectPressurePending = false
        clearWindows()
        controllerState = "cooldown"
        return pendingTargetKbps
    }

    fun recordApplied(targetKbps: Int) {
        val normalizedTarget = normalizeTarget(targetKbps)
        if (pendingTargetKbps > 0 && normalizedTarget != pendingTargetKbps) {
            return
        }
        val confirmsPendingTarget = normalizedTarget == pendingTargetKbps
        val confirmedBaseline = pendingBaselineTargetKbps?.takeIf { confirmsPendingTarget }
        if (confirmedBaseline != null) {
            baselineTargetKbps = confirmedBaseline
            floorTargetKbps = floorForBaseline(confirmedBaseline)
        }
        val minimumTargetKbps = listOfNotNull(
            floorTargetKbps,
            effectiveTargetKbps,
            baselineTargetKbps,
            pendingTargetKbps.takeIf { it > 0 }
        ).minOrNull() ?: floorTargetKbps
        val target = normalizedTarget.coerceIn(minimumTargetKbps, baselineTargetKbps)
        val appliedDecisionType = pendingDecisionType?.takeIf { confirmsPendingTarget }
        effectiveTargetKbps = target
        if (appliedDecisionType == "reduce") {
            automaticReductionCount += 1
        } else if (appliedDecisionType == "restore") {
            automaticRestorationCount += 1
        }
        pendingTargetKbps = 0
        pendingBaselineTargetKbps = null
        pendingDecisionType = null
        controllerState = if (effectiveTargetKbps < baselineTargetKbps) "reduced" else "observing"
    }

    fun recordRequested(targetKbps: Int) {
        pendingTargetKbps = normalizeTarget(targetKbps).coerceIn(floorTargetKbps, baselineTargetKbps)
        pendingBaselineTargetKbps = null
        pendingDecisionType = null
        clearWindows()
        controllerState = "cooldown"
    }

    fun recordFailure(reason: String, nowWallMs: Long) {
        pendingTargetKbps = 0
        pendingBaselineTargetKbps = null
        pendingDecisionType = null
        reconnectPressurePending = false
        clearWindows()
        controllerState = "failed"
        lastDecisionAt = nowWallMs.coerceAtLeast(lastDecisionAt)
        lastDecisionReason = reason.take(160)
    }

    fun evaluate(sample: NativeAdaptiveBitrateSample): NativeAdaptiveBitrateDecision? {
        if (controllerState == "failed") return null
        val nowElapsedMs = sample.nowElapsedMs.coerceAtLeast(0L)
        val nowWallMs = sample.nowWallMs.coerceAtLeast(0L)
        val droppedVideoFrames = sample.droppedVideoFrames.coerceAtLeast(0L)
        val droppedIncrease = (droppedVideoFrames - lastDroppedVideoFrames).coerceAtLeast(0L)
        val reconnectIncrease = (sample.cumulativeReconnectCount - cumulativeReconnectCount).coerceAtLeast(0)
        val thermalState = sample.thermalState.trim().lowercase()
        val thermalCritical = thermalState == "critical"
        val thermalSerious = thermalState == "serious"
        val thermalRecoveryBlocked = thermalState == "fair" || thermalSerious || thermalCritical
        val thermalFloorTargetKbps = floorForBaseline(pendingBaselineTargetKbps ?: baselineTargetKbps)
        val thermalSafetyTargetKbps = listOfNotNull(
            thermalFloorTargetKbps,
            effectiveTargetKbps.takeIf { it > 0 },
            pendingTargetKbps.takeIf { it > 0 }
        ).minOrNull() ?: thermalFloorTargetKbps
        if (reconnectIncrease > 0) {
            reconnectPressurePending = true
        }
        lastDroppedVideoFrames = droppedVideoFrames
        cumulativeReconnectCount = sample.cumulativeReconnectCount.coerceAtLeast(cumulativeReconnectCount)

        if (!sample.active) {
            clearWindows()
            controllerState = if (effectiveTargetKbps < baselineTargetKbps) "reduced" else "idle"
            return null
        }

        if (sample.publishGeneration != publishGeneration) {
            publishGeneration = sample.publishGeneration.coerceAtLeast(0)
            publishedAtElapsedMs = nowElapsedMs
            recoveryEligibleElapsedMs = nowElapsedMs + RECOVERY_AFTER_REPUBLISH_MS
            clearWindows()
            controllerState = "startup"
            if (!thermalCritical) return null
        }

        if (thermalCritical && pendingTargetKbps > 0 && pendingTargetKbps != thermalSafetyTargetKbps) {
            return decide(
                "reduce",
                thermalSafetyTargetKbps,
                pressureReason(sample, 0.0, 1.0, 0, 0),
                nowElapsedMs,
                nowWallMs,
                preservePendingBaseline = true
            )
        }

        if ((nowElapsedMs - publishedAtElapsedMs < STARTUP_GRACE_MS && !thermalCritical) || pendingTargetKbps > 0) {
            clearWindows()
            controllerState = if (pendingTargetKbps > 0) "cooldown" else "startup"
            return null
        }

        val queueRatio = queueRatio(sample.queuedItems, sample.cacheSize)
        val measuredRatio = if (sample.useMeasuredBitrate && sample.measuredBitrateKbps > 0 && effectiveTargetKbps > 0) {
            sample.measuredBitrateKbps.toDouble() / effectiveTargetKbps.toDouble()
        } else {
            1.0
        }
        val reconnectPressure = reconnectPressurePending
        val pressured = thermalSerious || thermalCritical || sample.congested || queueRatio >= 0.5 ||
            measuredRatio < 0.65 || droppedIncrease > 0 || reconnectPressure
        val critical = thermalSerious || thermalCritical || queueRatio >= 0.8 || measuredRatio < 0.4 ||
            droppedIncrease >= 3 || reconnectPressure

        if (pressured) {
            pressureSampleCount += 1
            criticalPressureSampleCount = if (critical) criticalPressureSampleCount + 1 else 0
            healthySampleCount = 0
            controllerState = "pressure"
            val holdReady = thermalCritical || pressureSampleCount >= PRESSURE_SAMPLES ||
                criticalPressureSampleCount >= CRITICAL_PRESSURE_SAMPLES
            val cooldownReady = thermalCritical || nowElapsedMs >= cooldownUntilElapsedMs
            val minimumTargetKbps = if (thermalCritical) thermalFloorTargetKbps else floorTargetKbps
            if (holdReady && cooldownReady && effectiveTargetKbps > minimumTargetKbps) {
                val target = if (thermalCritical) {
                    minimumTargetKbps
                } else {
                    maxOf(minimumTargetKbps, roundToHundred(effectiveTargetKbps * 0.8))
                }
                if (target < effectiveTargetKbps) {
                    return decide(
                        "reduce",
                        target,
                        pressureReason(sample, queueRatio, measuredRatio, droppedIncrease, if (reconnectPressure) 1 else 0),
                        nowElapsedMs,
                        nowWallMs,
                        preservePendingBaseline = thermalCritical
                    )
                }
            }
            if (holdReady && cooldownReady && effectiveTargetKbps <= minimumTargetKbps) {
                reconnectPressurePending = false
            }
            return null
        }

        clearPressure()
        val measuredHealthy = !sample.useMeasuredBitrate ||
            (sample.measuredBitrateKbps > 0 && measuredRatio >= 0.85)
        val healthy = effectiveTargetKbps < baselineTargetKbps && queueRatio <= 0.1 && measuredHealthy &&
            !thermalRecoveryBlocked
        if (!healthy) {
            healthySampleCount = 0
            controllerState = if (effectiveTargetKbps < baselineTargetKbps) "reduced" else "observing"
            return null
        }

        healthySampleCount += 1
        controllerState = "recovering"
        val recoveryReady = healthySampleCount >= HEALTHY_SAMPLES &&
            nowElapsedMs >= recoveryEligibleElapsedMs &&
            nowElapsedMs >= cooldownUntilElapsedMs
        if (!recoveryReady) return null
        val target = minOf(baselineTargetKbps, maxOf(effectiveTargetKbps + 100, roundToHundred(effectiveTargetKbps * 1.1)))
        return if (target > effectiveTargetKbps) {
            decide(
                "restore",
                target,
                "Native publisher remained uncongested through the recovery hold.",
                nowElapsedMs,
                nowWallMs
            )
        } else {
            null
        }
    }

    fun snapshot(nowElapsedMs: Long): NativeAdaptiveBitrateSnapshot = NativeAdaptiveBitrateSnapshot(
        controllerState = controllerState,
        baselineTargetKbps = baselineTargetKbps,
        effectiveTargetKbps = effectiveTargetKbps,
        floorTargetKbps = floorTargetKbps,
        pendingTargetKbps = pendingTargetKbps,
        automaticReductionCount = automaticReductionCount,
        automaticRestorationCount = automaticRestorationCount,
        pressureSampleCount = pressureSampleCount,
        healthySampleCount = healthySampleCount,
        cooldownRemainingMs = (cooldownUntilElapsedMs - nowElapsedMs).coerceAtLeast(0L),
        recoveryEligibleInMs = (recoveryEligibleElapsedMs - nowElapsedMs).coerceAtLeast(0L),
        publishGeneration = publishGeneration,
        cumulativeReconnectCount = cumulativeReconnectCount,
        lastDecisionAt = lastDecisionAt,
        lastDecisionReason = lastDecisionReason
    )

    private fun decide(
        type: String,
        targetKbps: Int,
        reason: String,
        nowElapsedMs: Long,
        nowWallMs: Long,
        preservePendingBaseline: Boolean = false
    ): NativeAdaptiveBitrateDecision {
        val retainedPendingBaseline = pendingBaselineTargetKbps.takeIf { preservePendingBaseline }
        pendingTargetKbps = targetKbps
        pendingBaselineTargetKbps = retainedPendingBaseline
        pendingDecisionType = type
        if (type == "reduce") {
            cooldownUntilElapsedMs = nowElapsedMs + REDUCTION_COOLDOWN_MS
        } else {
            cooldownUntilElapsedMs = nowElapsedMs + RECOVERY_COOLDOWN_MS
        }
        reconnectPressurePending = false
        lastDecisionAt = maxOf(lastDecisionAt, nowWallMs)
        lastDecisionReason = reason.take(160)
        clearWindows()
        controllerState = "cooldown"
        return NativeAdaptiveBitrateDecision(type, targetKbps, lastDecisionReason)
    }

    private fun clearWindows() {
        clearPressure()
        healthySampleCount = 0
    }

    private fun clearPressure() {
        pressureSampleCount = 0
        criticalPressureSampleCount = 0
    }

    private fun normalizeTarget(value: Int): Int = roundToHundred(value.coerceAtLeast(0).toDouble())

    private fun floorForBaseline(baselineKbps: Int): Int = minOf(
        baselineKbps,
        maxOf(ABSOLUTE_MINIMUM_KBPS, roundToHundred(baselineKbps * MINIMUM_BASELINE_PERCENT / 100.0))
    )

    private fun roundToHundred(value: Double): Int = ((value / 100.0).roundToInt() * 100).coerceAtLeast(0)

    private fun queueRatio(items: Int, capacity: Int): Double = if (capacity > 0) {
        (items.coerceAtLeast(0).toDouble() / capacity.toDouble()).coerceIn(0.0, 1.0)
    } else {
        0.0
    }

    private fun pressureReason(
        sample: NativeAdaptiveBitrateSample,
        queueRatio: Double,
        measuredRatio: Double,
        droppedIncrease: Long,
        reconnectIncrease: Int
    ): String = when {
        sample.thermalState.trim().equals("critical", ignoreCase = true) ->
            "The device reported critical thermal pressure; bitrate was reduced to the safety floor."
        sample.thermalState.trim().equals("serious", ignoreCase = true) ->
            "The device reported sustained serious thermal pressure."
        reconnectIncrease > 0 -> "A transport reconnect occurred during the active session."
        droppedIncrease > 0 -> "$droppedIncrease new publisher video frame drop(s) were observed."
        sample.congested || queueRatio >= 0.5 -> "Publisher queue pressure reached ${(queueRatio * 100).roundToInt()}%."
        measuredRatio < 0.65 -> "Measured send bitrate stayed at ${(measuredRatio * 100).roundToInt()}% of target."
        else -> "Sustained native publisher pressure was observed."
    }
}

File: ios/MobileLiveCasterBroadcastUpload/NativeAdaptiveBitrateController.swift
import Foundation

struct NativeAdaptiveBitrateSample {
    let nowElapsedMs: Double
    let nowWallMs: Double
    let active: Bool
    let publishGeneration: Int
    let congested: Bool
    let queuedItems: Int
    let cacheSize: Int
    let measuredBitrateKbps: Int
    let useMeasuredBitrate: Bool
    let droppedVideoFrames: Int
    let cumulativeReconnectCount: Int
    let thermalState: String
}

struct NativeAdaptiveBitrateDecision: Equatable {
    let type: String
    let targetKbps: Int
    let reason: String
}

struct NativeAdaptiveBitrateSnapshot: Equatable {
    let controlOwner: String
    let controllerState: String
    let baselineTargetKbps: Int
    let effectiveTargetKbps: Int
    let floorTargetKbps: Int
    let pendingTargetKbps: Int
    let automaticReductionCount: Int
    let automaticRestorationCount: Int
    let pressureSampleCount: Int
    let healthySampleCount: Int
    let cooldownRemainingMs: Int
    let recoveryEligibleInMs: Int
    let publishGeneration: Int
    let cumulativeReconnectCount: Int
    let lastDecisionAt: Double
    let lastDecisionReason: String

    func asDictionary() -> [String: Any] {
        [
            "controlOwner": controlOwner,
            "controllerState": controllerState,
            "baselineTargetKbps": baselineTargetKbps,
            "effectiveTargetKbps": effectiveTargetKbps,
            "floorTargetKbps": floorTargetKbps,
            "pendingTargetKbps": pendingTargetKbps,
            "automaticReductionCount": automaticReductionCount,
            "automaticRestorationCount": automaticRestorationCount,
            "pressureSampleCount": pressureSampleCount,
            "healthySampleCount": healthySampleCount,
            "cooldownRemainingMs": cooldownRemainingMs,
            "recoveryEligibleInMs": recoveryEligibleInMs,
            "publishGeneration": publishGeneration,
            "cumulativeReconnectCount": cumulativeReconnectCount,
            "lastDecisionAt": lastDecisionAt,
            "lastDecisionReason": lastDecisionReason
        ]
    }
}

struct NativeAdaptiveBitrateController {
    static let startupGraceMs = 10_000.0
    static let reductionCooldownMs = 8_000.0
    static let recoveryCooldownMs = 20_000.0
    static let recoveryAfterRepublishMs = 30_000.0
    static let pressureSamples = 3
    static let criticalPressureSamples = 2
    static let healthySamples = 30
    static let absoluteMinimumKbps = 900

    private var controllerState = "idle"
    private var baselineTargetKbps = 0
    private var effectiveTargetKbps = 0
    private var floorTargetKbps = 0
    private var pendingTargetKbps = 0
    private var pendingBaselineTargetKbps: Int?
    private var pendingDecisionType: String?
    private var automaticReductionCount = 0
    private var automaticRestorationCount = 0
    private var pressureSampleCount = 0
    private var criticalPressureSampleCount = 0
    private var healthySampleCount = 0
    private var cooldownUntilMs = 0.0
    private var recoveryEligibleMs = 0.0
    private var publishedAtMs = 0.0
    private var publishGeneration = 0
    private var cumulativeReconnectCount = 0
    private var reconnectPressurePending = false
    private var lastDroppedVideoFrames = 0
    private var lastDecisionAt = 0.0
    private var lastDecisionReason = ""

    mutating func reset(baselineKbps: Int, nowElapsedMs: Double = 0) {
        baselineTargetKbps = normalizeTarget(baselineKbps)
        effectiveTargetKbps = baselineTargetKbps
        floorTargetKbps = floorForBaseline(baselineTargetKbps)
        pendingTargetKbps = 0
        pendingBaselineTargetKbps = nil
        pendingDecisionType = nil
        automaticReductionCount = 0
        automaticRestorationCount = 0
        pressureSampleCount = 0
        criticalPressureSampleCount = 0
        healthySampleCount = 0
        cooldownUntilMs = 0
        recoveryEligibleMs = max(0, nowElapsedMs)
        publishedAtMs = 0
        publishGeneration = 0
        cumulativeReconnectCount = 0
        reconnectPressurePending = false
        lastDroppedVideoFrames = 0
        lastDecisionAt = 0
        lastDecisionReason = ""
        controllerState = "idle"
    }

    @discardableResult
    mutating func requestBaselineChange(_ baselineKbps: Int, thermalState: String = "unknown") -> Int {
        let target = normalizeTarget(baselineKbps)
        let normalizedThermalState = thermalState.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let currentTarget = [effectiveTargetKbps, pendingTargetKbps]
            .filter { $0 > 0 }
            .min() ?? target
        switch normalizedThermalState {
        case "critical":
            pendingTargetKbps = min(floorForBaseline(target), currentTarget)
        case "fair", "serious":
            pendingTargetKbps = min(target, currentTarget)
        default:
            pendingTargetKbps = target
        }
        pendingBaselineTargetKbps = target
        pendingDecisionType = nil
        reconnectPressurePending = false
        clearWindows()
        controllerState = "cooldown"
        return pendingTargetKbps
    }

    mutating func recordApplied(_ targetKbps: Int) {
        let normalizedTarget = normalizeTarget(targetKbps)
        if pendingTargetKbps > 0, normalizedTarget != pendingTargetKbps {
            return
        }
        let confirmsPendingTarget = normalizedTarget == pendingTargetKbps
        if confirmsPendingTarget, let confirmedBaseline = pendingBaselineTargetKbps {
            baselineTargetKbps = confirmedBaseline
            floorTargetKbps = floorForBaseline(confirmedBaseline)
        }
        let minimumTargetKbps = [floorTargetKbps, effectiveTargetKbps, baselineTargetKbps, pendingTargetKbps]
            .filter { $0 > 0 }
            .min() ?? floorTargetKbps
        let target = min(baselineTargetKbps, max(minimumTargetKbps, normalizedTarget))
        let appliedDecisionType = confirmsPendingTarget ? pendingDecisionType : nil
        effectiveTargetKbps = target
        if appliedDecisionType == "reduce" {
            automaticReductionCount += 1
        } else if appliedDecisionType == "restore" {
            automaticRestorationCount += 1
        }
        pendingTargetKbps = 0
        pendingBaselineTargetKbps = nil
        pendingDecisionType = nil
        controllerState = effectiveTargetKbps < baselineTargetKbps ? "reduced" : "observing"
    }

    mutating func recordFailure(_ reason: String, nowWallMs: Double) {
        pendingTargetKbps = 0
        pendingBaselineTargetKbps = nil
        pendingDecisionType = nil
        reconnectPressurePending = false
        clearWindows()
        controllerState = "failed"
        lastDecisionAt = max(lastDecisionAt, nowWallMs)
        lastDecisionReason = String(reason.prefix(160))
    }

    mutating func evaluate(_ sample: NativeAdaptiveBitrateSample) -> NativeAdaptiveBitrateDecision? {
        guard controllerState != "failed" else { return nil }
        let nowElapsedMs = max(0, sample.nowElapsedMs)
        let nowWallMs = max(0, sample.nowWallMs)
        let droppedVideoFrames = max(0, sample.droppedVideoFrames)
        let droppedIncrease = max(0, droppedVideoFrames - lastDroppedVideoFrames)
        let reconnectIncrease = max(0, sample.cumulativeReconnectCount - cumulativeReconnectCount)
        let thermalState = sample.thermalState.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let thermalCritical = thermalState == "critical"
        let thermalSerious = thermalState == "serious"
        let thermalRecoveryBlocked = thermalState == "fair" || thermalSerious || thermalCritical
        let thermalFloorTargetKbps = floorForBaseline(pendingBaselineTargetKbps ?? baselineTargetKbps)
        let thermalSafetyTargetKbps = [thermalFloorTargetKbps, effectiveTargetKbps, pendingTargetKbps]
            .filter { $0 > 0 }
            .min() ?? thermalFloorTargetKbps
        if reconnectIncrease > 0 {
            reconnectPressurePending = true
        }
        lastDroppedVideoFrames = droppedVideoFrames
        cumulativeReconnectCount = max(cumulativeReconnectCount, sample.cumulativeReconnectCount)

        guard sample.active else {
            clearWindows()
            controllerState = effectiveTargetKbps < baselineTargetKbps ? "reduced" : "idle"
            return nil
        }
        if sample.publishGeneration != publishGeneration {
            publishGeneration = max(0, sample.publishGeneration)
            publishedAtMs = nowElapsedMs
            recoveryEligibleMs = nowElapsedMs + Self.recoveryAfterRepublishMs
            clearWindows()
            controllerState = "startup"
            if !thermalCritical { return nil }
        }
        if thermalCritical, pendingTargetKbps > 0, pendingTargetKbps != thermalSafetyTargetKbps {
            return decide(
                "reduce",
                thermalSafetyTargetKbps,
                pressureReason(sample, 0, 1, 0, 0),
                nowElapsedMs,
                nowWallMs,
                preservePendingBaseline: true
            )
        }
        guard (nowElapsedMs - publishedAtMs >= Self.startupGraceMs || thermalCritical), pendingTargetKbps == 0 else {
            clearWindows()
            controllerState = pendingTargetKbps > 0 ? "cooldown" : "startup"
            return nil
        }

        let queueRatio = normalizedQueueRatio(items: sample.queuedItems, capacity: sample.cacheSize)
        let measuredRatio = sample.useMeasuredBitrate && sample.measuredBitrateKbps > 0 && effectiveTargetKbps > 0
            ? Double(sample.measuredBitrateKbps) / Double(effectiveTargetKbps)
            : 1
        let reconnectPressure = reconnectPressurePending
        let pressured = thermalSerious || thermalCritical || sample.congested || queueRatio >= 0.5 ||
            measuredRatio < 0.65 || droppedIncrease > 0 || reconnectPressure
        let critical = thermalSerious || thermalCritical || queueRatio >= 0.8 || measuredRatio < 0.4 ||
            droppedIncrease >= 3 || reconnectPressure
        if pressured {
            pressureSampleCount += 1
            criticalPressureSampleCount = critical ? criticalPressureSampleCount + 1 : 0
            healthySampleCount = 0
            controllerState = "pressure"
            let holdReady = thermalCritical || pressureSampleCount >= Self.pressureSamples ||
                criticalPressureSampleCount >= Self.criticalPressureSamples
            let minimumTargetKbps = thermalCritical ? thermalFloorTargetKbps : floorTargetKbps
            if holdReady, (thermalCritical || nowElapsedMs >= cooldownUntilMs), effectiveTargetKbps > minimumTargetKbps {
                let target = thermalCritical
                    ? minimumTargetKbps
                    : max(minimumTargetKbps, roundToHundred(Double(effectiveTargetKbps) * 0.8))
                if target < effectiveTargetKbps {
                    return decide(
                        "reduce",
                        target,
                        pressureReason(sample, queueRatio, measuredRatio, droppedIncrease, reconnectPressure ? 1 : 0),
                        nowElapsedMs,
                        nowWallMs,
                        preservePendingBaseline: thermalCritical
                    )
                }
            }
            if holdReady, (thermalCritical || nowElapsedMs >= cooldownUntilMs), effectiveTargetKbps <= minimumTargetKbps {
                reconnectPressurePending = false
            }
            return nil
        }

        clearPressure()
        let measuredHealthy = !sample.useMeasuredBitrate ||
            (sample.measuredBitrateKbps > 0 && measuredRatio >= 0.85)
        guard effectiveTargetKbps < baselineTargetKbps, queueRatio <= 0.1, measuredHealthy, !thermalRecoveryBlocked else {
            healthySampleCount = 0
            controllerState = effectiveTargetKbps < baselineTargetKbps ? "reduced" : "observing"
            return nil
        }
        healthySampleCount += 1
        controllerState = "recovering"
        guard healthySampleCount >= Self.healthySamples, nowElapsedMs >= recoveryEligibleMs, nowElapsedMs >= cooldownUntilMs else {
            return nil
        }
        let target = min(baselineTargetKbps, max(effectiveTargetKbps + 100, roundToHundred(Double(effectiveTargetKbps) * 1.1)))
        return target > effectiveTargetKbps
            ? decide(
                "restore",
                target,
                "Native publisher remained uncongested through the recovery hold.",
                nowElapsedMs,
                nowWallMs
            )
            : nil
    }

    func snapshot(nowElapsedMs: Double) -> NativeAdaptiveBitrateSnapshot {
        NativeAdaptiveBitrateSnapshot(
            controlOwner: "native",
            controllerState: controllerState,
            baselineTargetKbps: baselineTargetKbps,
            effectiveTargetKbps: effectiveTargetKbps,
            floorTargetKbps: floorTargetKbps,
            pendingTargetKbps: pendingTargetKbps,
            automaticReductionCount: automaticReductionCount,
            automaticRestorationCount: automaticRestorationCount,
            pressureSampleCount: pressureSampleCount,
            healthySampleCount: healthySampleCount,
            cooldownRemainingMs: max(0, Int(cooldownUntilMs - nowElapsedMs)),
            recoveryEligibleInMs: max(0, Int(recoveryEligibleMs - nowElapsedMs)),
            publishGeneration: publishGeneration,
            cumulativeReconnectCount: cumulativeReconnectCount,
            lastDecisionAt: lastDecisionAt,
            lastDecisionReason: lastDecisionReason
        )
    }

    private mutating func decide(
        _ type: String,
        _ targetKbps: Int,
        _ reason: String,
        _ nowElapsedMs: Double,
        _ nowWallMs: Double,
        preservePendingBaseline: Bool = false
    ) -> NativeAdaptiveBitrateDecision {
        let retainedPendingBaseline = preservePendingBaseline ? pendingBaselineTargetKbps : nil
        pendingTargetKbps = targetKbps
        pendingBaselineTargetKbps = retainedPendingBaseline
        pendingDecisionType = type
        if type == "reduce" {
            cooldownUntilMs = nowElapsedMs + Self.reductionCooldownMs
        } else {
            cooldownUntilMs = nowElapsedMs + Self.recoveryCooldownMs
        }
        reconnectPressurePending = false
        lastDecisionAt = max(lastDecisionAt, nowWallMs)
        lastDecisionReason = String(reason.prefix(160))
        clearWindows()
        controllerState = "cooldown"
        return NativeAdaptiveBitrateDecision(type: type, targetKbps: targetKbps, reason: lastDecisionReason)
    }

    private mutating func clearWindows() {
        clearPressure()
        healthySampleCount = 0
    }

    private mutating func clearPressure() {
        pressureSampleCount = 0
        criticalPressureSampleCount = 0
    }

    private func normalizeTarget(_ value: Int) -> Int { roundToHundred(Double(max(0, value))) }
    private func floorForBaseline(_ baseline: Int) -> Int {
        min(baseline, max(Self.absoluteMinimumKbps, roundToHundred(Double(baseline) * 0.35)))
    }
    private func roundToHundred(_ value: Double) -> Int { max(0, Int((value / 100).rounded()) * 100) }
    private func normalizedQueueRatio(items: Int, capacity: Int) -> Double {
        capacity > 0 ? min(1, max(0, Double(items) / Double(capacity))) : 0
    }
    private func pressureReason(
        _ sample: NativeAdaptiveBitrateSample,
        _ queueRatio: Double,
        _ measuredRatio: Double,
        _ droppedIncrease: Int,
        _ reconnectIncrease: Int
    ) -> String {
        if sample.thermalState.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == "critical" {
            return "The device reported critical thermal pressure; bitrate was reduced to the safety floor."
        }
        if sample.thermalState.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == "serious" {
            return "The device reported sustained serious thermal pressure."
        }
        if reconnectIncrease > 0 { return "A transport reconnect occurred during the active session." }
        if droppedIncrease > 0 { return "\(droppedIncrease) new publisher video frame drop(s) were observed." }
        if sample.congested || queueRatio >= 0.5 { return "Publisher queue pressure reached \(Int((queueRatio * 100).rounded()))%." }
        if measuredRatio < 0.65 { return "Measured send bitrate stayed at \(Int((measuredRatio * 100).rounded()))% of target." }
        return "Sustained native publisher pressure was observed."
    }
}
